addons/survey/controllers/main.py

```diff
@@ -710,55 +710,90 @@ def _generate_report(self, user_input, download=True):
+                        question = answers[0].question_id
+                        answer_domain = expression.AND([
+                            ['&', ('value_char_box', '=ilike', value), ('question_id', '=', question.id)],
+                            answer_domain
+                        ])
+
+                    if question_type != 'char_box':
+                        value = '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
+                    if answer_domain:
+                        filters_domain.append(answer_domain)
+                    filters_data.append({
+                        'question_type': question_type,
                         'row_id': row_id,
                         'answer_id': answer_id,
-                        'question': question_id.title,
-                        'answers': '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
+                        'question_id': question.id,
+                        'question': question.title,
+                        'answers': value
                     })
-        if line_choices:
-            line_filter_domain = expression.AND([[('suggested_answer_id', 'in', line_choices)], line_filter_domain])
 
-        user_input_domain = self._get_user_input_domain(survey, line_filter_domain, **post)
-        user_input_lines = request.env['survey.user_input'].sudo().search(user_input_domain).mapped('user_input_line_ids')
+        # Get the survey user inputs
+        user_input_domain = ['&', ('test_entry', '=', False), ('survey_id', '=', survey.id)]
+        if post.get('finished'):
+            user_input_domain = expression.AND([[('state', '=', 'done')], user_input_domain])
+        else:
+            user_input_domain = expression.AND([[('state', '!=', 'new')], user_input_domain])
+        if post.get('failed'):
+            user_input_domain = expression.AND([[('scoring_success', '=', False)], user_input_domain])
+        elif post.get('passed'):
+            user_input_domain = expression.AND([[('scoring_success', '=', True)], user_input_domain])
+        user_inputs = request.env['survey.user_input'].sudo().search(user_input_domain)
+
+        # Filter the user inputs
+        filtered_user_input = request.env['survey.user_input']
+        for domain in filters_domain:
+            recordset = user_inputs.mapped('user_input_line_ids').filtered_domain(domain).mapped('user_input_id')
+            if not filtered_user_input:
+                filtered_user_input = recordset
+            else:
+                filtered_user_input &= recordset
+
+        for suggested_answer_id in suggested_answer_ids:
+            recordset = user_inputs.mapped('user_input_line_ids').filtered_domain([('suggested_answer_id', '=', suggested_answer_id)]).mapped('user_input_id')
+            if not filtered_user_input:
+                filtered_user_input = recordset
+            else:
+                filtered_user_input &= recordset
```

Comment:
To avoid making mapped + filter domain + re- mapped (and make multiple search at the end, because everything - related records attributes - is not loaded in cache with the first search), I propose to use a search read with the famous big fat domain.
Then, we have all we need to process it in python. A bit more code but only one call to DB.
I made an example with suggested_answer_id (the easiest part :smile: ) and it work quite fine.
```
        # Create the big fat domain to collect all the user_input_lines matching the filters independently
        input_line_domain = ['&', ('user_input_id', 'in', user_inputs.ids), ('suggested_answer_id', 'in', suggested_answer_ids)]
        for domain in filters_domain:
            input_line_domain = expression.OR([input_line_domain, domain])
        results = request.env['survey.user_input.line'].search_read(
            input_line_domain,
            fields=['id', 'user_input_id', 'answer_type', 'suggested_answer_id', 'value_char_box', 'matrix_row_id']
        )
        
        # Then, Keep only the user_input_lines that match all the filters
        from collections import defaultdict # To put at start of file
        input_line_by_answer_id = defaultdict(list)
        for result in results:
            # Filtered Domain
            if not result['suggested_answer_id']:
                continue # TODO
            # Suggested Answer Id
            else:
                input_line_by_answer_id[result['suggested_answer_id'][0]].append(result['user_input_id'][0])

        filtered_user_input_ids = set.intersection(*map(set,input_line_by_answer_id.values()))
```